Upgrade AdSense Connect CTA to new design.

techanvil opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #5260 

## Relevant technical choices

This PR should not be merged until the placeholder images have been updated. I've therefore created it as a Draft, but moved into Code Review so we can get started on the review process and address anything that comes out of it while waiting for the final images.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

asvinb commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/index.js:
```diff
@@ -124,89 +130,61 @@ export default function AdSenseConnectCTA() {
-			global.removeEventListener( 'keyup', handleDialogClose );
-		};
-	}, [] );
+	const cellProps = {
```
@techanvil From a readability perspective, I would prefer if we inline those props because when looking at a specific `Cell`, we immediately know its size, instead of going back to the variable and check.

techanvil replied:
@asvinb although I see your point to an extent, the problem with inlining the props here it is obscures the fact that the cells do in fact share these props, and it's necessary to compare them manually in order to notice that. To me, that makes it a bit less readable as it makes this higher level aspect, that the cells are "the same", harder to spot.

asvinb reviewed (changes requested):
@techanvil This is a good start. Please look at my comments and let me know what you think. For the carousel/slider feature, I'am wondering whether something like tiny slider would have been best suited here. Also I spent some time diving into a CSS only solution using `scroll-snap` but had some troubles.

techanvil commented on assets/sass/components/adsense/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,191 @@
+
+	.googlesitekit-setup__wrapper--adsense-connect {
+
+		.googlesitekit-setup-module--adsense__stage-caption-container {
```
Yeah, good thinking, that's done :+1:

techanvil commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/Content.js:
```diff
@@ -0,0 +1,152 @@
+					</h2>
+				</div>
+			</Cell>
+			<Cell { ...cellProps }></Cell>
```
This is to enable the image cell to span the grid row in non-mobile viewports in order to achieve the desired visual layout. I've added a couple of inline comments to explain this.

asvinb replied:
@techanvil Check my comment here: https://github.com/google/site-kit-wp/pull/5336#discussion_r905228794 😁

techanvil commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/Content.js:
```diff
@@ -0,0 +1,152 @@
+	};
+
+	return (
+		<Row ref={ ref }>
```
This is also related to the row-spanning of the image cell (at non-mobile viewports). As the `Row` component is actually a `display: grid` element, it's not possible to span cells across multiple `Row` components. Tbh I was thinking of doing without `Grid/Row/Cell` here, but I found it useful to reuse some of the existing `Cell` props.

Maybe there's a better way of achieving the layout here - this does seems OK to me, but I am open to suggestion...

techanvil commented on assets/sass/components/adsense/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,191 @@
+
+		@media (min-width: $bp-desktop) {
+
+			.googlesitekit-setup-module__action {
```
Sounds good. I also had to add `flex-wrap: wrap` in order to allow the buttons to wrap to two rows at very narrow viewports.

techanvil commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/ContentSwipeable.js:
```diff
@@ -0,0 +1,48 @@
+const minStage = 0;
+const maxStage = 2;
+
+export default function ContentSwipeable() {
```
Yeah, I actually wondered this myself, I was waiting to get some feedback as the idea of which way is "correct" is somewhat subjective (and it's not specified in the AC). Seeing as you also feel the inverse would be more natural, I've gone ahead and made that change :+1:

techanvil commented on assets/sass/components/adsense/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,191 @@
+			margin: 0;
+			opacity: 0;
+
+			transition-duration: 2.5s;
```
I think, as this is quite clearly specified in the AC, we should leave this, as going from 2.5s to 0.5s is a significant reduction. My preference is to implement as specced and we can easily tweak it based on downstream feedback.

asvinb replied:
Sounds good. Let's follow up on that. Especially on mobile, it looks so slow and does not feel natural 👍

techanvil commented:
> @techanvil This is a good start. Please look at my comments and let me know what you think. For the carousel/slider feature, I'am wondering whether something like tiny slider would have been best suited here. Also I spent some time diving into a CSS only solution using `scroll-snap` but had some troubles.

Thanks for the review @asvinb, I've addressed your comments.

As to the usage of something like tiny slider, my opinion is that the requirements are probably bespoke enough to warrant our own implementation without using a library. For example the fact that the images and text are treated differently to each other at the various breakpoints (images always overlap & fade in/out while text only does so at non-mobile breakpoints), and the responsive layout complicates things further, I imagine we would need to run the library code twice (similarly to how we currently render the images & text separately in two loops), possibly re-configuring at the breakpoint change, and I expect we would need to customise the CSS a fair amount anyway. Also there's the addition to the bundle size to consider when adding a library - even tiny slider adds 31.5kb according to https://bundlephobia.com/package/tiny-slider@2.9.4 which although quite small would still be nice to avoid if we don't actually need it. So, as stated I think it's OK as it is, but if you have a clear idea in mind that we could improve things via using a library I'd be interested to hear more about it.

As to the idea of a CSS-only solution, I did discuss this with @aaemnnosttv but we came to the conclusion that it's not really necessary to pursue this for the sake of CSS purity, rather stick with something that should be easier to maintain even if it's using a bit of JS where we may be able to do without it.

aaemnnosttv commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/ContentSVG.js:
```diff
@@ -0,0 +1,64 @@
+/**
```
Let's keep the lazy loading for now. The AdSense JS is loaded almost all the time but these graphics will only be shown until dismissed (or connected) so it's best to not include them all the time. Also, these graphics could change in size later at which point a large change in bundle size could be missed.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+.googlesitekit-plugin {
+
+	.googlesitekit-setup__wrapper--adsense-connect {
+		$bp-nonMobile: $width-tablet + 1 + px;
```
@techanvil Very nice. I can see usage of that breakpoint elsewhere. How about we move it to `assets/sass/config/_variables.scss`

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+			grid-row-start: 1;
+			margin: 0;
+			opacity: 0;
+
```
@techanvil Nitpick but let's remove the empty line.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+		}
+
+		.googlesitekit-setup-module--adsense__stage-caption {
+
```
@techanvil Nitpick but let's remove the empty line.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+				font-size: 18px;
+				font-weight: 400;
+				margin: 0 0 4px;
+
```
@techanvil Nitpick but let's remove the empty line.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+			p {
+				font-size: 12px;
+				margin: 0;
+
```
@techanvil Nitpick but let's remove the empty line.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+				transition: inherit;
+			}
+
+			@media (min-width: $bp-nonMobile) {
```
@techanvil Let's move this before the styles for `h4`, [L50](https://github.com/google/site-kit-wp/pull/5336/files#diff-49489c5b6c975530fa4dfea0af9ceae9fae06785debf85cff2a98c299293c717R50)

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+				color: $c-secondary;
+			}
+
+			@media (min-width: $bp-nonMobile) {
```
@techanvil Same, let's move this just before the `h4` styles for the selector.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+				max-width: 100%;
+			}
+
+			@media (min-width: $bp-nonMobile) {
```
@techanvil Move just before `svg` styles.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+				}
+			}
+
+			@media (min-width: $bp-nonMobile) {
```
Same here, to move before the `li` styles.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+				margin: 0;
+			}
+
+			@media (min-width: $bp-nonMobile) {
```
@techanvil Same here.

asvinb commented on assets/sass/components/setup/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,186 @@
+			}
+		}
+
+		@media (min-width: $bp-nonMobile) {
```
@techanvil Let's avoid wrapping selectors within breakpoints :-)

asvinb commented on assets/sass/components/adsense/_googlesitekit-adsense-connect-cta.scss:
```diff
@@ -0,0 +1,191 @@
+			}
+		}
+
+		@media (min-width: $width-tablet + 1 + px) {
```
That's fine with the nested @media rules. It's easier to read. I've noticed there are a couple of places still left, which i've pointed out :)

techanvil commented:
Thanks @asvinb, I've addressed your comments. Thanks too for taking the time to tackle the layout issue, I think your negative margin is a great improvement and have cherry-picked the commit as-is.

However - I have also made further changes. @aaemnnosttv pointed out that the animation has been fleshed out in an [design prototype](https://www.figma.com/proto/jdJaDjJijEzsuFH5rhpZvo/AdSense-Advert?page-id=10%3A3&node-id=888%3A2184&viewport=672%2C-570%2C0.21&scaling=min-zoom&starting-point-node-id=888%3A2184&hotspot-hints=0) and I had completely failed to take this on board, although it's linked in the ACs. So, I've made some changes to reflect the animation on display in the prototype. This has made things a bit more complicated, but that is par for the course I suppose. Please see what you think!

tofumatt commented on .storybook/preview-head-vrt.html:
```diff
@@ -6,4 +6,11 @@
 		transition-duration: 0ms !important;
 	}
+
+	#root .googlesitekit-vrt-animation-none *,
+	#root .googlesitekit-vrt-animation-none *::before,
+	#root .googlesitekit-vrt-animation-none *::after {
+		animation: none !important;
+		transition: none !important;
+	}
```
These are somewhat obscure selectors with important modifiers. I'm not opposed to `!important`, but my usual rule around using it is to document _why_, so the next person who encounters it doesn't think "oh, is this even needed?" and then remove it or wonder what it's for 😅 

Can you add a comment here explaining what this does?

tofumatt commented on assets/js/components/Tooltip.js:
```diff
@@ -0,0 +1,81 @@
+			spotlightPadding={ 0 }
+			floaterProps={ floaterProps }
+			locale={ joyrideLocale }
+			run={ true }
```
```suggestion
			run
```

I wouldn't thought we had an ESLint rule for boolean attributes that would've flagged this 🤔

techanvil replied:
That would be useful :). I have removed the `={ true }` and added a followup ESLint issue - https://github.com/google/site-kit-wp/issues/5508

tofumatt commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/index.test.js:
```diff
@@ -52,57 +47,60 @@ describe( 'AdSenseConnectCTA', () => {
+			container = render(
+				<div>
+					<div id="adminmenu">
+						<a href="?page=googlesitekit-settings">Settings</a>
```
This is minor, but this link stood out to me as odd when browsing the code. Is there any reason not to make it a full URL? It'd be more accurate anyway to what we'd encounter in production, and looks less weird at a glance 😅 

```suggestion
						<a href="http://test.test/wp-admin/admin.php?page=googlesitekit-settings">Settings</a>
```

techanvil replied:
No reason not to at all, I have amended as per your suggestion!

tofumatt commented on assets/js/modules/adsense/components/common/AdSenseConnectCTA/index.test.js:
```diff
@@ -52,57 +47,60 @@ describe( 'AdSenseConnectCTA', () => {
+				</div>,
+				{ registry }
+			).container;
+			// await act( async () => {
```
```suggestion

```

tofumatt reviewed (changes requested):
Looks like the medium and large stories are missing the SVG image. They look quite odd, I guess it's a bug?

<img width="1020" alt="CleanShot 2022-06-29 at 18 35 55" src="https://user-images.githubusercontent.com/90871/176500330-6a0b1025-7b45-4084-99a3-94cf5eea7f90.png">

-----

A few other things:

* The fading animation (fade-out/fade-in) is quite slow. I'd say we can increase it a fair bit because it just seems painfully slow to me.
* Especially with the long animation time, a user can encounter a weird state if they scroll down to Monetization after about 5 seconds:

![CleanShot 2022-06-29 at 18 49 20](https://user-images.githubusercontent.com/90871/176503064-ac102fe3-a079-4b61-a340-be009aba9145.gif)

We probably shouldn't start the animation only once the user has viewed the slideshow, or they'll miss the first section and need to wait for it to loop back. Or at least increase the animation speed so it isn't possible to scroll down and see almost nothing in the image space for more than a second.

The fade-in on page load, especially with the slow speed, is a bit odd too:

![CleanShot 2022-06-29 at 18 52 44](https://user-images.githubusercontent.com/90871/176503422-4755871e-0bd1-4e18-b1aa-8e48626afa79.gif)

Again, with a much faster animation it shouldn't be as bad, but I'd expect that image to just be immediately loaded on page load, _then_ animate after that. I guess if it's lazy-loaded maybe we can't do that? 🤔 

The speed is really noticeable when swiping on mobile. It makes the animation feel very laggy:

![CleanShot 2022-06-29 at 18 54 28](https://user-images.githubusercontent.com/90871/176503788-0f0d7bd4-3b07-4564-b641-9361763693cb.gif)

So let's make that animation way faster. It should be in the hundreds of milliseconds, not seconds, I think. But feel free to experiment.

Overall looks good though, aside from the points in this summary mostly small nitpicks 🙂

aaemnnosttv commented:
@techanvil @tofumatt Moving this back to develop for the next release as this likely won't make it through QA before release QA has started and this is a substantial change.

techanvil commented:
Thanks for the review @tofumatt!

> Looks like the medium and large stories are missing the SVG image. They look quite odd, I guess it's a bug?

It was deliberate at the time, due to the way the image faded in at the start, but (following up on another point of yours) I've now updated the component so the image is fully visible at the start (notwithstanding the lazy load). The VRT screengrabs are now captured with the image visible, a definite improvement.

> * The fading animation (fade-out/fade-in) is quite slow. I'd say we can increase it a fair bit because it just seems painfully slow to me.

So, I haven't changed the animation speed in the non-mobile viewport because it's a pretty close match to the [design prototype](https://www.figma.com/proto/jdJaDjJijEzsuFH5rhpZvo/AdSense-Advert?page-id=10%3A3&node-id=888%3A2184&viewport=672%2C-570%2C0.21&scaling=min-zoom&starting-point-node-id=888%3A2184&hotspot-hints=0) linked in the AC. If you time it you'll see they are roughly the same.

However, I _have_ gone ahead and increased the animation speed in the mobile viewport, because this has not been fleshed out in the prototype and as you point out it was pretty sluggy feeling, especially as it's a response to user input in the mobile view whereas the non-mobile is an auto-update.

> * Especially with the long animation time, a user can encounter a weird state if they scroll down to Monetization after about 5 seconds:

I've addressed this issue by a) not fading in the first image, and b) not starting the animation until the component is in view.

> We probably shouldn't start the animation only once the user has viewed the slideshow, or they'll miss the first section and need to wait for it to loop back. Or at least increase the animation speed so it isn't possible to scroll down and see almost nothing in the image space for more than a second.

Yep, great shout and I have addressed as mentioned above.

> The fade-in on page load, especially with the slow speed, is a bit odd too:
> 
> Again, with a much faster animation it shouldn't be as bad, but I'd expect that image to just be immediately loaded on page load, _then_ animate after that. I guess if it's lazy-loaded maybe we can't do that? thinking

Likewise, this one.

> The speed is really noticeable when swiping on mobile. It makes the animation feel very laggy:
> So let's make that animation way faster. It should be in the hundreds of milliseconds, not seconds, I think. But feel free to experiment.

And this too!

I've addressed your other comments as well, so sending back to you for another look. Cheers

tofumatt reviewed (changes requested):
The mobile speed here is still very slow-feeling:

![CleanShot 2022-07-06 at 12 29 50](https://user-images.githubusercontent.com/90871/177540506-b060b228-93fb-4cdc-90a4-515e0e0eda71.gif)

I think we either entirely or almost-entirely nix the fade-in/fade-out animation here. A "scroll" type animation on mobile would be much more appropriate here, it's something we should file a follow-up issue for and do after this merges.

But I think the long animation, especially where there's like a second of nearly entirely blank image on screen—because the new image doesn't start fading in until the old one is gone—is just much too long. I wonder if the Figma mock-up was made slower to illustrate the animation? At any rate I think it's just _much_ too slow and we need to decrease it. It's distracting and makes the site feel slow.

![CleanShot 2022-07-06 at 12 31 53](https://user-images.githubusercontent.com/90871/177540848-abcee06f-0f94-4b1d-8fa0-969b6527e017.gif)

Generally speaking, animations should signify something and often happen in response to user activity. That this one doesn't isn't _great_, but at least it should be shorter.

I'm going to experiment with the speed and ask for a merge review, but after that we can send this to QA I think 👍🏻

tofumatt commented:
I've adjusted the animation speed to be a lot slower, but mostly to avoid long periods of blank images:

![CleanShot 2022-07-06 at 13 21 26](https://user-images.githubusercontent.com/90871/177549098-0bfef4cc-171e-4c63-898a-bc89abed6cf1.gif)

I've also adjusted it on mobile:

![CleanShot 2022-07-06 at 13 21 00](https://user-images.githubusercontent.com/90871/177549075-5fec1235-311a-4a39-b7d9-cf86a0441ba7.gif)

tofumatt commented:
Note: the colours are a bit off in that GIF on non-mobile screens. It actually looks like this:

https://user-images.githubusercontent.com/90871/177550021-a3fb8f34-a703-42c0-ab19-66e4c3ce9d71.mp4

tofumatt commented:
Chatted to @techanvil and this looks good to him so we're merging. 👍🏻

tofumatt reviewed (approved):
Changes made, so good-to-go 🥳
